### PR TITLE
fix(harvest): order between file with the same dep tech

### LINKED ISSUE #13: Bad sorting techs in levels
```js
const builder = Builder({
    levels: [
        'bem/libs/bem-core/common.blocks',
        'bem/libs/bem-core/desktop.blocks',
        'bem/libs/bem-components/common.blocks',
        'bem/libs/bem-components/desktop.blocks',
        'bem/libs/bem-components/design/common.blocks',
        'bem/libs/bem-components/design/desktop.blocks',
        'bem/libs/bem-forms/common.blocks'
    ],
    techMap: {
        js: ['vanilla.js', 'source.js', 'browser.js', 'js'],
        css: ['styl', 'post.css', 'css']
    }
});
```

Exapmle block: __checkbox__

In level 'bem-forms' tech.name __checkbox.browser.js__
In level 'bem-components' tech.name __checkbox.js__

Durning build __checkbox.js__ added after __checkbox.browser.js__

### PR DESCRIPTION
**Pull request by qfox**
Fixes #13

### CODE REVIEW

**Comment by qfox** [comment]
cc @voischev

**Comment by qfox** [comment]
cc @blond @belozer @Yeti-or @tadatuta COMEON LETS MERGE THIS

File: index.js [hunk 1]
```diff
@@ -199,40 +196,42 @@ function filesToStream(filesPromise, options) {
 }
 
 /**
- * @param {Array<{entity: BemEntityName, level: String, tech: String, path: String}>} introspection - unordered file-entities list
- * @param {String[]} levels - ordered levels' paths list
- * @param {Tenorok[]} decl - resolved and ordered declaration
+ * @param {Object} opts - Options for harvester
+ * @param {Array<{entity: BemEntityName, level: String, tech: String, path: String}>} opts.introspection - unordered file-entities list
```
Comment (Yeti-or) [inline index.js:200]: <BemFile>

File: index.js [hunk 2]
```diff
@@ -199,40 +196,42 @@ function filesToStream(filesPromise, options) {
...
-        .filter(file => levels.indexOf(file.level) !== -1)
-        .sort((f1, f2) => hash(f1) === hash(f2)
-            ? levels.indexOf(f1.level) - levels.indexOf(f2.level)
-            : declIndex[hash(f1)] - declIndex[hash(f2)]);
+function harvest(opts) {
+    const hash = cell => `${cell.entity.id}.${cell.tech}`;
+    const declIndex = _buildIndex(opts.decl, hash);
+
+    const fileTechToDep = Object.keys(opts.techMap || {}).reduce((res, depTech) =>
+        (opts.techMap[depTech].forEach(fileTech => (res[fileTech] = depTech)), res),
```
Comment (Yeti-or) [inline index.js]: comma operator, really?

File: index.js [hunk 3]
```diff
@@ -199,40 +196,42 @@ function filesToStream(filesPromise, options) {
...
+ * @param {BemCell[]} list - List of tenoroks
  * @param {Function} hash - Hashing function
  * @returns {Object<String, Number>} - Entity id to sort order
  */
 function _buildIndex(list, hash) {
-    return list.reduce((res, fileEntity, idx) => {
-        res[hash(fileEntity)] = idx;
+    return list.reduce((res, cell, idx) => {
+        res[cell.entity.id] = idx;
+        res[hash(cell)] = idx;
```
Comment (Yeti-or) [inline index.js]: why you need to provide `hash` function
it could be different?
could we move `hash` out of `harvest` as helper?

File: index.js [hunk 4]
```diff
@@ -199,40 +196,42 @@ function filesToStream(filesPromise, options) {
...
+    const declIndex = _buildIndex(opts.decl, hash);
+
+    const fileTechToDep = Object.keys(opts.techMap || {}).reduce((res, depTech) =>
+        (opts.techMap[depTech].forEach(fileTech => (res[fileTech] = depTech)), res),
+        {});
+
+    return opts.introspection
+        .filter(file => {
+            return declIndex[hash(file)] !== undefined ||
+                fileTechToDep[file.tech] && declIndex[hash(Object.assign({}, file, {tech: fileTechToDep[file.tech]}))] !== undefined;
```
Comment (Yeti-or) [inline index.js]: could you add some comments about this code ?

File: index.js [hunk 5]
```diff
@@ -199,40 +196,42 @@ function filesToStream(filesPromise, options) {
...
-        .filter(entityInIndex)
-        .filter(file => levels.indexOf(file.level) !== -1)
-        .sort((f1, f2) => hash(f1) === hash(f2)
-            ? levels.indexOf(f1.level) - levels.indexOf(f2.level)
-            : declIndex[hash(f1)] - declIndex[hash(f2)]);
+function harvest(opts) {
+    const hash = cell => `${cell.entity.id}.${cell.tech}`;
+    const declIndex = _buildIndex(opts.decl, hash);
+
+    const fileTechToDep = Object.keys(opts.techMap || {}).reduce((res, depTech) =>
```
Comment (Yeti-or) [inline index.js]: extToTech ?

File: index.js (hunk 1 above)
Comment (qfox) [reply index.js:200]: Ye, here should be a BemFile

File: index.js (hunk 5 above)
Comment (qfox) [reply index.js]: Ext is not correct.

```
           ↓↓ - ext
block.post.js
      ^^^^^^^ - tech
```

**Comment by Yeti-or** [comment]
https://github.com/gulp-bem/gulp-bem-src/pull/18

File: index.js (hunk 2 above)
Comment (qfox) [reply index.js]: You don't like it? ;-(

File: index.js (hunk 4 above)
Comment (qfox) [reply index.js]: Well, I guess we should refactor this code. Need more tests and use `@bem/cell` before refactoring.
Atm it just filter out the introspection before sorting according to passed in options (decl, levels, techMap).

File: index.js (hunk 5 above)
Comment (Yeti-or) [reply index.js]: so techMap is `ext` to `tech`  ?

File: index.js (hunk 3 above)
Comment (qfox) [reply index.js]: Well, prob we don't need it at all. We have cell.id and entity.id for these purposes. Also would be nice to have `cell.partialEquality` and `cell.mutate` functions.

File: index.js (hunk 2 above)
Comment (Yeti-or) [reply index.js]: I like it, really but you know this is not so maintainable

File: index.js (hunk 5 above)
Comment (qfox) [reply index.js]: dependency tech to file system tech.

`{ block: 'button', tech: 'react.js' }` by default should look for `**/button/button.react.js` (roughly).
But with techMap you can force it to look for, let's say, `**/button/button.component.js` or some else.

It's definitely not an extension.

File: index.js (hunk 2 above)
Comment (qfox) [reply index.js]: We will rewrite this part of code a lot of times soon.

**Pull request merged**